x-pack/plugins/endpoint/server/routes/resolver/utils/pagination.ts

```diff
@@ -56,36 +63,45 @@ export function getPaginationParams(limit: number, after?: string): PaginationPa
   return { size: limit };
 }
 
-export function paginate(pagination: PaginationParams, field: string, query: JsonObject) {
+export function paginate(
+  pagination: PaginationParams,
+  tiebreaker: string,
+  aggregator: string,
+  query: JsonObject
+): JsonObject {
   const { size, timestamp, eventID } = pagination;
-  query.sort = [{ '@timestamp': 'asc' }, { [field]: 'asc' }];
-  query.aggs = { total: { value_count: { field } } };
+  query.sort = [{ '@timestamp': 'asc' }, { [tiebreaker]: 'asc' }];
+  query.aggs = query.aggs || {};
+  query.aggs = Object.assign({}, query.aggs, { totals: { terms: { field: aggregator, size } } });
   query.size = size;
   if (timestamp && eventID) {
     query.search_after = [timestamp, eventID] as Array<number | string>;
   }
   return query;
 }
 
-export function paginatedResults(
-  response: SearchResponse<ResolverEvent>
-): { total: number; results: ResolverEvent[]; nextCursor: string | null } {
-  const total = response.aggregations?.total?.value || 0;
-  if (response.hits.hits.length === 0) {
-    return { total, results: [], nextCursor: null };
+export function buildPaginationCursor(total: number, results: ResolverEvent[]): string | null {
+  if (total > results.length && results.length > 0) {
+    const lastResult = results[results.length - 1];
+    const cursor = {
+      timestamp: lastResult['@timestamp'],
+      eventID: entityId(lastResult),
+    };
+    return urlEncodeCursor(cursor);
   }
+  return null;
+}
 
-  const results: ResolverEvent[] = [];
-  for (const hit of response.hits.hits) {
-    results.push(hit._source);
+export function paginatedResults(response: SearchResponse<ResolverEvent>): PaginatedResults {
+  if (response.hits.hits.length === 0) {
+    return { totals: {}, results: [] };
   }
 
-  // results will be at least 1 because of length check at the top of the function
-  const next = results[results.length - 1];
-  const cursor = {
-    timestamp: next['@timestamp'],
-    eventID: extractEventID(next),
-  };
+  const totals = response.aggregations?.totals?.buckets?.reduce(
```

Comment:
❔  In the `d.ts` I see `aggregations` typed as `any` - Sure it's safe to assume `buckets` is `reduce`able?

Reply:
I think the response is structured based on the type of query you pass to ES. I wonder if there's a way to check the structure of the response before running the `reduce`? It'd probably be an ES error if the response didn't come back the way it should, or we messed up the query.